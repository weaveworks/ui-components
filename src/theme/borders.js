import { kebabCase, forEach } from 'lodash';

export const borderRadius = {
  circle: '50%',
  none: '0',
  soft: '2px',
};

// Collects all theme border radii as SCSS vars.
export function themeBorderRadiiAsScss() {
  const themeBorderRadii = [];

  forEach(borderRadius, (value, name) => {
    themeBorderRadii.push(`$border-radius-${kebabCase(name)}: ${value}`);
  });

  return themeBorderRadii;
}
