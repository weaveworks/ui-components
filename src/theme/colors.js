import { isString, kebabCase, forEach } from 'lodash';

export const colors = {
  // Neutral Colors
  black: 'hsl(0, 0%, 10%)', // #1a1a1a
  blue50: 'hsl(191, 100%, 97%)', // #f0fcff
  blue200: 'hsl(191, 100%, 80%)', // #99ecff
  blue400: 'hsl(191, 100%, 50%)', // #00d2ff
  blue600: 'hsl(191, 100%, 40%)', // #00a7cc
  blue700: 'hsl(191, 100%, 35%)', // #0092b3
  blue800: 'hsl(191, 100%, 30%)', // #007d99
  graphThemes: {
    blue: [
      '#c7e9b4',
      '#7ecdbb',
      '#1eb5eb',
      '#1d91bf',
      '#235fa9',
      '#253393',
      '#084181',
    ],
    mixed: [
      '#c7e9b4',
      '#c1d4e7',
      '#7ecdbb',
      '#9fbddb',
      '#1eb5eb',
      '#8d95c6',
      '#1d91bf',
      '#8282ab',
      '#235fa9',
      '#89429e',
      '#253393',
      '#800f7a',
      '#084181',
      '#0b0533',
    ],
    purple: [
      '#c1d4e7',
      '#9fbddb',
      '#8d95c6',
      '#8282ab',
      '#89429e',
      '#800f7a',
      '#0b0533',
    ],
  }, // Used by PrometheusGraph component
  gray50: 'hsl(0, 0%, 96%)', // #f4f4f4
  gray100: 'hsl(0, 0%, 90%)', // #e6e6e6
  gray200: 'hsl(0, 0%, 80%)', // #cccccc

  // #737373
  gray600: 'hsl(0, 0%, 45%)',
  green500: 'hsl(161, 54%, 48%)', // #38bd93
  orange500: 'hsl(13, 100%, 50%)', // #ff3700
  orange600: 'hsl(13, 100%, 40%)', // #cc2c00
  orange700: 'hsl(13, 100%, 35%)', // #b32700
  // Accent Colors
  // #992100
  orange800: 'hsl(13, 100%, 30%)',
  promQL: {
    attrName: '#00a4db',
    /**
     * GHColors theme by Avi Aryan (http://aviaryan.in)
     * Inspired by Github syntax coloring
     */
    comment: '#bbbbbb',
    deleted: '#9a050f',
    entity: '#36acaa',
    function: '#dc322f',
    metricName: '#2aa198',
    punctuation: '#393a34',
    // Dropdown colors,
    salmon: '#ff7c7c',
    string: '#e3116c',
    tag: '#00009f',
  }, // PromQL
  purple25: 'hsl(240, 20%, 98%)', // #fafafc
  purple50: 'hsl(240, 20%, 95%)', // #eeeef4
  purple100: 'hsl(240, 20%, 90%)', // #dfdfea
  purple200: 'hsl(240, 20%, 75%)', // #b1b1cb
  purple300: 'hsl(240, 20%, 65%)', // #9494b8
  purple400: 'hsl(240, 20%, 60%)', // #8585ad

  purple500: 'hsl(240, 20%, 50%)', // #666699
  purple600: 'hsl(240, 20%, 45%)', // #5b5b88
  purple700: 'hsl(240, 20%, 35%)', // #47476b
  purple800: 'hsl(240, 20%, 30%)', // #3d3d5c
  // Primary Colors
  // #32324b
  purple900: 'hsl(240, 20%, 25%)',

  // Third-party specific colors - not to be used in the theme!
  thirdParty: {
    azure: '#3769bb',
    // Google single-click login
    cornflowerBlue: '#4285f4',
  },
  // #ffffff
  white: 'hsl(0, 0%, 100%)',
  // #d4ab27
  yellow500: 'hsl(46, 69%, 49%)',
};

// Flattens and collects all theme colors as SCSS vars
export function themeColorsAsScss() {
  const ignoreKeys = ['graphThemes'];
  const themeColors = [];

  forEach(colors, (value, name) => {
    const colorPrefix = `$color-${kebabCase(name)}`;
    if (ignoreKeys.includes(name)) return;

    if (isString(value)) {
      themeColors.push(`${colorPrefix}: ${value}`);
    } else {
      forEach(value, (innerValue, subname) => {
        if (ignoreKeys.includes(subname)) return;
        themeColors.push(`${colorPrefix}-${kebabCase(subname)}: ${innerValue}`);
      });
    }
  });

  return themeColors;
}
