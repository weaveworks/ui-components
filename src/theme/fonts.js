import { kebabCase, forEach } from 'lodash';

export const fontFamilies = {
  monospace: "'Roboto Mono', monospace",
  regular: "'proxima-nova', Helvetica, Arial, sans-serif",
};

export const fontSizes = {
  extraLarge: '32px',
  huge: '48px',
  large: '22px',
  normal: '16px',
  small: '14px',
  tiny: '12px',
};

// Collects all theme font vars as SCSS vars.
export function themeFontsAsScss() {
  const themeFonts = [];

  forEach(fontFamilies, (value, name) => {
    themeFonts.push(`$font-family-${kebabCase(name)}: ${value}`);
  });

  forEach(fontSizes, (value, name) => {
    themeFonts.push(`$font-size-${kebabCase(name)}: ${value}`);
  });

  return themeFonts;
}
