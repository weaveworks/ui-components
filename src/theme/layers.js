import { kebabCase, forEach } from 'lodash';

export const layers = {
  alert: 3,
  dropdown: 5,
  front: 1,
  modal: 7,
  notification: 4,
  toolbar: 2,
  tooltip: 6,
};

// Collects all theme z-index layers as SCSS vars
export function themeLayersAsScss() {
  const themeLayers = [];

  forEach(layers, (value, name) => {
    themeLayers.push(`$layer-${kebabCase(name)}: ${value}`);
  });

  return themeLayers;
}
