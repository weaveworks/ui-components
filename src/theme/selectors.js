import { get, isObject } from 'lodash';

// `selector()` allows you to create closures over as
// many levels as you wish, each time returning another function.
// This lets you compose selectors and when they are passed an object
// finally select that value
export const selector = (propsOrSelector, prevArgs = []) => {
  if (isObject(propsOrSelector)) {
    return get(propsOrSelector, prevArgs);
  }

  return value => selector(value, [...prevArgs, propsOrSelector]);
};

const themeSelector = selector('theme');

// these are pre-composed selectors for our theme
// Usage:
//   const Success = styled.i`
//     color: ${color('green500')};
//   `;
export const boxShadow = themeSelector('boxShadow');
export const borderRadius = themeSelector('borderRadius');
export const color = themeSelector('colors');
export const fontSize = themeSelector('fontSizes');
