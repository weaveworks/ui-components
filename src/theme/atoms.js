import { colors } from './colors';

// component-specific
export const atoms = {
  Alert: {
    error: {
      background: colors.orange600,
      color: colors.white,
    },
    info: {
      background: colors.blue600,
      color: colors.white,
    },
    success: {
      background: colors.green500,
      color: colors.white,
    },
    warning: {
      background: colors.yellow500,
      color: colors.white,
    },
  },
  Button: {
    danger: {
      background: colors.orange600,
      color: colors.white,
      hoverBackground: colors.orange700,
      hoverColor: colors.white,
    },
    default: {
      background: colors.white,
      color: colors.black,
      hoverBackground: colors.gray50,
      hoverColor: colors.purple800,
    },
    disabled: {
      background: colors.gray50,
      color: colors.gray600,
      hoverBackground: colors.gray50,
      hoverColor: colors.gray600,
    },
    primary: {
      background: colors.blue700,
      color: colors.white,
      hoverBackground: colors.blue800,
      hoverColor: colors.white,
    },
  },
};
