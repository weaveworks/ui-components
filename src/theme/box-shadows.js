import { colors } from './colors';

export const boxShadow = {
  heavy: '0 10px 20px rgba(0,0,0,0.19), 0 6px 6px rgba(0,0,0,0.23)',
  light: '0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24)',
  none: 'none',
  selected: `0px 0px 2px 2px ${colors.purple400}`,
};
