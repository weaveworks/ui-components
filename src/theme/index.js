import { colors, themeColorsAsScss } from './colors';
import { fontSizes, fontFamilies, themeFontsAsScss } from './fonts';
import { borderRadius, themeBorderRadiiAsScss } from './borders';
import { layers, themeLayersAsScss } from './layers';
import { spacing, themeSpacingsAsScss } from './spacings';
import { overlayIconSize, textColor, themeMiscVarsAsScss } from './misc';
import { boxShadow } from './box-shadows';
import { atoms } from './atoms';

export function themeVarsAsScss() {
  const themeVariables = []
    .concat(themeColorsAsScss())
    .concat(themeLayersAsScss())
    .concat(themeFontsAsScss())
    .concat(themeBorderRadiiAsScss())
    .concat(themeSpacingsAsScss())
    .concat(themeMiscVarsAsScss());
  return `${themeVariables.join('; ')};`;
}

const theme = {
  // Component-specific
  atoms,

  // Borders
  borderRadius,

  // Box shadows
  boxShadow,

  // Colors
  colors,

  // Fonts
  fontFamilies,
  fontSizes,

  // z-index layers
  layers,

  // Misc
  overlayIconSize,

  // Spacings
  spacing,
  textColor,
};

export default theme;
