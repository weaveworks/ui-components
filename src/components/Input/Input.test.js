import React from 'react';
import { mount } from 'enzyme';
import 'jest-styled-components';

import { withTheme } from '../../utils/theme';

import Input from './Input';

describe('<Input />', () => {
  let props;
  let wrapper;
  let onChange;

  beforeEach(() => {
    onChange = jest.fn();

    props = {
      label: 'Hi',
      message: 'Doh!',
      onChange,
      valid: false,
    };

    wrapper = mount(withTheme(<Input {...props} />));
  });

  it('should call onChange on input', () => {
    expect(onChange).not.toHaveBeenCalled();

    wrapper.find('input').simulate('change');

    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it('should not render validation message when valid=true', () => {
    props = {
      message: 'Derp!',
      valid: true,
    };
    wrapper = mount(withTheme(<Input {...props} />), {
      lifecycleExperimental: true,
    });

    expect(wrapper.find('Input__ValidationMessage')).toHaveStyleRule(
      'visibility',
      'hidden'
    );
  });

  it('should render validation message when valid=false', () => {
    props = {
      message: 'Derp!',
      valid: false,
    };
    wrapper = mount(withTheme(<Input {...props} />));

    expect(wrapper.find('Input__ValidationMessage')).toHaveStyleRule(
      'visibility',
      'visible'
    );
  });

  it('should only render <label> if label prop is provided', () => {
    props = {
      message: 'Derp!',
      valid: false,
    };
    wrapper = mount(withTheme(<Input {...props} />));

    expect(wrapper.find('label').length).toBe(0);

    wrapper = mount(
      withTheme(<Input {...{ ...props, label: 'Woohoo, label time!' }} />)
    );

    expect(wrapper.find('label').length).toBe(1);
  });

  it('should remove validation message from the DOM when hideValidationMessage=true', () => {
    props = {
      hideValidationMessage: true,
    };
    wrapper = mount(withTheme(<Input {...props} />));

    expect(wrapper.find('Input__ValidationMessage')).toHaveStyleRule(
      'display',
      'none'
    );
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
