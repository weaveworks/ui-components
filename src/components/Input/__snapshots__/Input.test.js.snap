// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render correctly 1`] = `
.c2 {
  position: absolute;
  right: 10px;
  visibility: visible;
  color: hsl(13,100%,40%);
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 36px;
  margin-top: 8px;
}

.c1 input {
  padding-right: 32px;
  padding-left: 12px;
  width: 100%;
  line-height: 36px;
  box-shadow: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid hsl(0,0%,45%);
  border-radius: 2px;
  height: 36px;
  box-sizing: border-box;
}

.c1 input::-webkit-input-placeholder {
  color: hsl(0,0%,45%);
}

.c1 input::-moz-placeholder {
  color: hsl(0,0%,45%);
}

.c1 input:-moz-placeholder {
  color: hsl(0,0%,45%);
}

.c1 input:-ms-input-placeholder {
  color: hsl(0,0%,45%);
}

.c1 input::-webkit-input-placeholder {
  opacity: 1;
}

.c1 input::-moz-placeholder {
  opacity: 1;
}

.c1 input:-moz-placeholder {
  opacity: 1;
}

.c1 input:-ms-input-placeholder {
  opacity: 1;
}

.c1 input:disabled {
  background-color: hsl(0,0%,96%);
}

.c1 input:disabled::-webkit-input-placeholder {
  opacity: 0.5;
}

.c1 input:disabled::-moz-placeholder {
  opacity: 0.5;
}

.c1 input:disabled:-moz-placeholder {
  opacity: 0.5;
}

.c1 input:disabled:-ms-input-placeholder {
  opacity: 0.5;
}

.c3 {
  display: block;
  padding-left: 8px;
  margin-top: 8px;
  font-size: 14px;
  text-align: left;
  color: hsl(13,100%,40%);
  visibility: visible;
}

.c0 {
  padding: 8px;
}

<Component
  theme={
    Object {
      "atoms": Object {
        "Alert": Object {
          "error": Object {
            "background": "hsl(13, 100%, 40%)",
            "color": "hsl(0, 0%, 100%)",
          },
          "info": Object {
            "background": "hsl(191, 100%, 40%)",
            "color": "hsl(0, 0%, 100%)",
          },
          "success": Object {
            "background": "hsl(161, 54%, 48%)",
            "color": "hsl(0, 0%, 100%)",
          },
          "warning": Object {
            "background": "hsl(46, 69%, 49%)",
            "color": "hsl(0, 0%, 100%)",
          },
        },
        "Button": Object {
          "danger": Object {
            "background": "hsl(13, 100%, 40%)",
            "color": "hsl(0, 0%, 100%)",
            "hoverBackground": "hsl(13, 100%, 35%)",
            "hoverColor": "hsl(0, 0%, 100%)",
          },
          "default": Object {
            "background": "hsl(0, 0%, 100%)",
            "color": "hsl(0, 0%, 10%)",
            "hoverBackground": "hsl(0, 0%, 96%)",
            "hoverColor": "hsl(240, 20%, 30%)",
          },
          "disabled": Object {
            "background": "hsl(0, 0%, 96%)",
            "color": "hsl(0, 0%, 45%)",
            "hoverBackground": "hsl(0, 0%, 96%)",
            "hoverColor": "hsl(0, 0%, 45%)",
          },
          "primary": Object {
            "background": "hsl(191, 100%, 35%)",
            "color": "hsl(0, 0%, 100%)",
            "hoverBackground": "hsl(191, 100%, 30%)",
            "hoverColor": "hsl(0, 0%, 100%)",
          },
        },
      },
      "borderRadius": Object {
        "circle": "50%",
        "none": "0",
        "soft": "2px",
      },
      "boxShadow": Object {
        "heavy": "0 10px 20px rgba(0,0,0,0.19), 0 6px 6px rgba(0,0,0,0.23)",
        "light": "0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24)",
        "none": "none",
        "selected": "0px 0px 2px 2px hsl(240, 20%, 60%)",
      },
      "colors": Object {
        "black": "hsl(0, 0%, 10%)",
        "blue200": "hsl(191, 100%, 80%)",
        "blue400": "hsl(191, 100%, 50%)",
        "blue50": "hsl(191, 100%, 97%)",
        "blue600": "hsl(191, 100%, 40%)",
        "blue700": "hsl(191, 100%, 35%)",
        "blue800": "hsl(191, 100%, 30%)",
        "graphThemes": Object {
          "blue": Array [
            "#c7e9b4",
            "#7ecdbb",
            "#1eb5eb",
            "#1d91bf",
            "#235fa9",
            "#253393",
            "#084181",
          ],
          "mixed": Array [
            "#c7e9b4",
            "#c1d4e7",
            "#7ecdbb",
            "#9fbddb",
            "#1eb5eb",
            "#8d95c6",
            "#1d91bf",
            "#8282ab",
            "#235fa9",
            "#89429e",
            "#253393",
            "#800f7a",
            "#084181",
            "#0b0533",
          ],
          "purple": Array [
            "#c1d4e7",
            "#9fbddb",
            "#8d95c6",
            "#8282ab",
            "#89429e",
            "#800f7a",
            "#0b0533",
          ],
        },
        "gray100": "hsl(0, 0%, 90%)",
        "gray200": "hsl(0, 0%, 80%)",
        "gray50": "hsl(0, 0%, 96%)",
        "gray600": "hsl(0, 0%, 45%)",
        "green500": "hsl(161, 54%, 48%)",
        "orange500": "hsl(13, 100%, 50%)",
        "orange600": "hsl(13, 100%, 40%)",
        "orange700": "hsl(13, 100%, 35%)",
        "orange800": "hsl(13, 100%, 30%)",
        "promQL": Object {
          "attrName": "#00a4db",
          "comment": "#bbbbbb",
          "deleted": "#9a050f",
          "entity": "#36acaa",
          "function": "#dc322f",
          "metricName": "#2aa198",
          "punctuation": "#393a34",
          "salmon": "#ff7c7c",
          "string": "#e3116c",
          "tag": "#00009f",
        },
        "purple100": "hsl(240, 20%, 90%)",
        "purple200": "hsl(240, 20%, 75%)",
        "purple25": "hsl(240, 20%, 98%)",
        "purple300": "hsl(240, 20%, 65%)",
        "purple400": "hsl(240, 20%, 60%)",
        "purple50": "hsl(240, 20%, 95%)",
        "purple500": "hsl(240, 20%, 50%)",
        "purple600": "hsl(240, 20%, 45%)",
        "purple700": "hsl(240, 20%, 35%)",
        "purple800": "hsl(240, 20%, 30%)",
        "purple900": "hsl(240, 20%, 25%)",
        "thirdParty": Object {
          "azure": "#3769bb",
          "cornflowerBlue": "#4285f4",
        },
        "white": "hsl(0, 0%, 100%)",
        "yellow500": "hsl(46, 69%, 49%)",
      },
      "fontFamilies": Object {
        "monospace": "'Roboto Mono', monospace",
        "regular": "'proxima-nova', Helvetica, Arial, sans-serif",
      },
      "fontSizes": Object {
        "extraLarge": "32px",
        "huge": "48px",
        "large": "22px",
        "normal": "16px",
        "small": "14px",
        "tiny": "12px",
      },
      "layers": Object {
        "alert": 3,
        "dropdown": 5,
        "front": 1,
        "modal": 7,
        "notification": 4,
        "toolbar": 2,
        "tooltip": 6,
      },
      "overlayIconSize": "300px",
      "spacing": Object {
        "base": "16px",
        "large": "32px",
        "medium": "24px",
        "none": "0",
        "small": "12px",
        "xl": "48px",
        "xs": "8px",
        "xxl": "64px",
        "xxs": "4px",
      },
      "textColor": "hsl(0, 0%, 10%)",
    }
  }
>
  <Input
    label="Hi"
    message="Doh!"
    onChange={[MockFunction]}
    valid={false}
  >
    <Input
      autoSelectText={false}
      className="c0"
      focus={false}
      hideValidationMessage={false}
      inputRef={[Function]}
      label="Hi"
      message="Doh!"
      onChange={[MockFunction]}
      textarea={false}
      valid={false}
    >
      <div
        className="c0"
      >
        <label>
          Hi
        </label>
        <Input__InputWrapper
          hasLabel={true}
          valid={false}
        >
          <div
            className="c1"
          >
            <input
              onChange={[MockFunction]}
            />
            <Input__Icon
              className="fa fa-times-circle"
              visible={true}
            >
              <i
                className="c2 fa fa-times-circle"
              />
            </Input__Icon>
          </div>
        </Input__InputWrapper>
        <Input__ValidationMessage
          remove={false}
          valid={false}
          visible={true}
        >
          <span
            className="c3"
          >
            Doh!
          </span>
        </Input__ValidationMessage>
      </div>
    </Input>
  </Input>
</Component>
`;
